src/bootstrap/service.rs

```diff
@@ -0,0 +1,1285 @@
+        // They are flushed and the DB compacted only when it is closed.
+        // If we do nothing and write continuously, the writes slow down A LOT, bringing the replication to a crawl.
+        // So, as a work around, every 5M messages, we close and reopen the DB. This flushes and finializes all the keys written
+        // so far, saves progress, and creates a proper "checkpoint" for us to resume, in case things go wrong or the user
+        // presses CTRL+C etc...
+        let max_messages_in_write_session = 5_000_000;
+
+        tokio::spawn(async move {
+            info!("DB writer thread started for shard {}", shard_id);
+            let mut total_messages = 0;
+            let total_start_time = std::time::Instant::now();
+
+            while let Some(commit_request) = db_commit_rx.recv().await {
+                let start_time = std::time::Instant::now();
+                total_messages += commit_request.num_messages;
+
+                let result = db.commit(commit_request.txn_batch);
+                let elapsed_time = start_time.elapsed();
+                let total_elapsed = total_start_time.elapsed();
+                info!(
+                    "DB writer({}) - Commit took {:?}. Writing at {:.1}k msg/sec {:.2}M total messages",
+                    shard_id,
+                    elapsed_time,
+                    total_messages as f64 / 1000.0 / total_elapsed.as_secs_f64(),
+                    total_messages as f64 / 1_000_000.0
+                );
+
+                let response = match result {
+                    Ok(()) => {
+                        // If we've exceeded the total number of messages in the write session, stop the writer
+                        // and allow the DB to reload
+                        if total_messages > max_messages_in_write_session {
+                            Ok(DbCommitResponse::Stop)
+                        } else {
+                            Ok(DbCommitResponse::Continue)
+                        }
+                    }
+                    Err(e) => {
+                        error!("DB writer - Failed to commit transaction: {}", e);
+                        Err(BootstrapError::DatabaseError(format!(
+                            "Failed to commit transaction: {}",
+                            e
+                        )))
+                    }
+                };
+
+                // Send the result back to the requesting thread
+                if let Err(_) = commit_request.response_tx.send(response) {
+                    error!("Failed to send DB commit response from DB Writer thread");
+                    break;
+                }
+            }
+
+            info!("DB writer thread finished for shard {}", shard_id);
+        })
+    }
+
+    /// Create a dedicated thread (per shard) to write to the Trie. This way, we can keep all the trie nodes
```

Comment:
Is there churn in what's kept in memory if there's only 1 thread because the trie nodes get kicked out of memory when the account stores are written?

Reply:
So the trie does not get kicked out of memory after the account stores are written, because this thread keeps its own copy of the trie in memory. This is where most of the perf gain comes from. By keeping most of the top layers of the trie in memory, we can avoid upto 12 DB reads per insert, speeding up things quite a bit. This trie is reloaded after the db batch of 5 million messages. 

In any case, the main reasons to do a trie in its own thread is so that we don't have to do post-processing anymore. The trie root hash is always valid since there are no competing threads. I'll update the comment to reflect this.